### PR TITLE
 Feat/cli progress bar

### PR DESCRIPTION
**Pull request by dryajov**
this includes changes from #994

### CODE REVIEW

**Comment by dryajov** [comment]
This PR changes the behavior of `files.add` considerably as it switches to a chunked response (stream of objects) that reports progress. One issue with this, is that the response is streamed and it's not possible to report an error with a standard 500 code, since it's already been initiated with a 200 code. My current solution is to report an error as part of the stream of objects and catch it in the `ProgressStream` through stream, if an error object is detected (`{Message: '', Code: #}`) the stream emits the error, which should propagate correctly. 

I'm not sure if this is the best approach and would be curious to know how Go handles it. Tagging @whyrusleeping and @Kubuxu for possible input.

**Comment by whyrusleeping** [comment]
In go-ipfs we use an http trailer for errors

**Comment by dryajov** [comment]
@whyrusleeping ah! Didn't know about those - but makes sense! Thanks!

**Comment by dryajov** [comment]
Coverage has gone down due to https://github.com/ipfs/js-ipfs/pull/1036/files#diff-dcaf16a1df33e0e8f40195e527b9294dR244, but I don't see any tests that exercise that functionality directly. @diasdavid any ideas of how to tests this so that coverage goes up?

File: src/core/components/files.js [hunk 1]
```diff
@@ -65,6 +65,14 @@ module.exports = function files (self) {
...
+      let prog = options.progress || (() => {})
+      const progress = (bytes) => {
+        total += bytes
+        prog(total)
+      }
+
+      options.progress = progress
       pull(
         pull.values(normalizeContent(data)),
         importer(self._ipldResolver, options),
```
Comment (dryajov) [inline src/core/components/files.js]: @diasdavid @VictorBjelkholm I noticed that there are two entry points for importer, here and above on line 26. The one above, gets called from the HTTP API, this one is called from the regular callback flow, they seem to be be doing exactly the same thing, perhaps we can replace this with a call to `createAddPullStream` and consolidate the progress accumulator logic there?

File: package.json [hunk 2]
```diff
@@ -74,7 +74,7 @@
     "expose-loader": "^0.7.3",
     "form-data": "^2.3.1",
     "gulp": "^3.9.1",
-    "interface-ipfs-core": "~0.31.19",
+    "interface-ipfs-core": "^0.32.1",
```
Comment (diasdavid) [inline package.json]: @dryajov do not change ~ to ^, we use ~ until it gets to 1.0.0

**Comment by diasdavid** [comment]
@dryajov I've tested this locally and the progress bar is not perfect. You can see it showing but it reaches quickly 100% and disappears and then it takes 5 or more seconds to actually get the hash of the file printed.

I'm going to go ahead and merge as this has an improvement, but it should not be considered complete.

**Pull request merged**